File: cascadingFailuresTaskMigration_python/evaluation/evaluation_etra_target.py
import math
from typing import List
from input.robot import Robot
from input.task import Task


class EvaluationEtraTarget:
    def calculate_robot_capacity_std(self, robots: List[Robot]) -> float:
        """Calculate standard deviation of robot capacities."""
        capacity_sum = sum(robot.capacity for robot in robots)
        mean = capacity_sum / len(robots)
        sum_sqr = sum((robot.capacity - mean) ** 2 for robot in robots)
        return math.sqrt(sum_sqr / len(robots))

    def calculate_task_size_std(self, tasks: List[Task]) -> float:
        """Calculate standard deviation of task sizes."""
        size_sum = sum(task.size for task in tasks)
        mean = size_sum / len(tasks)
        sum_sqr = sum((task.size - mean) ** 2 for task in tasks)
        return math.sqrt(sum_sqr / len(tasks))

    def calculate_mean_robot_capacity(self, robots: List[Robot]) -> float:
        """Calculate mean robot capacity."""
        capacity_sum = sum(robot.capacity for robot in robots)
        return capacity_sum / len(robots)

    def calculate_mean_task_size(self, tasks: List[Task]) -> float:
        """Calculate mean task size."""
        size_sum = sum(task.size for task in tasks)
        return size_sum / len(tasks)

File: cascadingFailuresTaskMigration/src/main/java/evaluation/EvaluationEtraTarget.java
package evaluation;

import input.Robot;
import input.Task;

import java.util.List;

public class EvaluationEtraTarget {
    public Double calculateRobotCapacityStd(List<Robot> robots) {
        double capacitySum  = 0.0;
        for (Robot robot : robots) {
            capacitySum +=robot.getCapacity();
        }
        double mean = capacitySum /robots.size();
        double sumSqr = 0.0;
        for (Robot robot : robots) {
            double capacity = robot.getCapacity();
            sumSqr+=(capacity-mean)*(capacity-mean);
        }
        return Math.sqrt(sumSqr/robots.size());
    }

    public Double calculateTaskSizeStd(List<Task> tasks) {
        double sizeSum = 0.0;
        for (Task task : tasks) {
            sizeSum+=task.getSize();
        }
        double mean = sizeSum/tasks.size();
        double sumSqr = 0.0;
        for (Task task : tasks) {
            double size = task.getSize();
            sumSqr+=(size-mean)*(size-mean);
        }
        return Math.sqrt(sumSqr/tasks.size());
    }

    public Double calculateMeanRobotCapacity(List<Robot> robots) {
        double capacitySum = 0.0;
        for (Robot robot : robots) {
            capacitySum+=robot.getCapacity();
        }
        return capacitySum/robots.size();
    }

    public Double calculateMeanTaskSize (List<Task> tasks){
        double sizeSum = 0.0;
        for (Task task : tasks) {
            sizeSum+=task.getSize();
        }
        return sizeSum/tasks.size();
    }

}
